packages/graphql-connection-transformer/src/ModelConnectionTransformer.ts

```diff
@@ -46,6 +63,74 @@ function validateKeyField(field: FieldDefinitionNode): void {
     throw new InvalidDirectiveError(`If you define a field and specify it as a 'keyField', it must be of type 'ID' or 'String'.`)
 }
 
+/**
+ * Ensure that the field passed in is compatible to be a key field
+ * (Not a list and of type ID or String)
+ * @param field: the field to be checked.
+ */
+function validateKeyFieldNewConnection(field: FieldDefinitionNode): void {
+    if (!field) {
+        return
+    }
+    const isNonNull = isNonNullType(field.type);
+    const isAList = isListType(field.type)
+
+    // The only valid key fields are single non-null fields.
+    if (!isAList && isNonNull) {
+        return;
+    }
+    throw new InvalidDirectiveError(`All fields used for a connection cannot be lists and must be of Non-Null types.`)
+}
+
+/**
+ * Checks that the fields being used to query match the expected key types for the index being used.
+ * @param parentFields: All fields of the parent object.
+ * @param realtedTypeFields: All fields of the related object.
+ * @param inputFieldNames: The fields passed in to the @connection directive.
+ * @param keySchema: The key schema for the index being used.
+ */
+function checkFieldsAgainstIndex(parentFields: ReadonlyArray<FieldDefinitionNode>,
+                                 relatedTypeFields: ReadonlyArray<FieldDefinitionNode>,
+                                 inputFieldNames: string[],
+                                 keySchema: KeySchema[]): void {
+    const hashAttributeName = keySchema[0].AttributeName;
+    const tablePKType = relatedTypeFields.find(f => f.name.value === hashAttributeName).type;
+    const queryPKType = parentFields.find(f => f.name.value === inputFieldNames[0]).type;
+    const numFields = inputFieldNames.length;
+
+    if (getBaseType(tablePKType) !== getBaseType(queryPKType)) {
+        throw new InvalidDirectiveError(inputFieldNames[0] + ' field is not of type ' + getBaseType(tablePKType))
+    }
+    if (numFields > keySchema.length && keySchema.length !== 2) {
+        throw new InvalidDirectiveError('Too many fields passed in to @connection directive.')
+    }
+    if (numFields === 2) {
```

Comment:
What about if numFields === 1?

Reply:
Then there is only a partition key provided and that case is already checked with the first if-condition on line 101.